### PR TITLE
Unify and embed multiple filters that operate on the same property

### PR DESCRIPTION
**Pull request by andrejtonev**
Introduced new operator RangeOperator

Rewriting the Cypher AST tree to combine multiple filters into a range filter
Inserting range filter into range scans

### CODE REVIEW

File: src/query/frontend/ast/pretty_print.cpp [hunk 1]
```diff
@@ -251,6 +252,13 @@ BINARY_OPERATOR_VISIT(SubscriptOperator, "Subscript");
 
 #undef BINARY_OPERATOR_VISIT
 
+void ExpressionPrettyPrinter::Visit(RangeOperator &op) {
```
Comment (Ignition) [inline src/query/frontend/ast/pretty_print.cpp:255]: How do we reach this code/see its effect?

File: src/query/plan/preprocess.cpp [hunk 2]
```diff
@@ -184,7 +186,8 @@ void AddExpansionsToMatching(std::vector<Expansion> &expansions, Matching &match
   }
 }
 
-auto SplitExpressionOnAnd(Expression *expression) {
+template <typename F1, typename F2>
+void SplitExpressionOnAnd(Expression *expression, F1 &&if_cb, F2 &&else_cb) {
```
Comment (Ignition) [inline src/query/plan/preprocess.cpp:190]: Only two usages...inline into both places and refactor them to have simple implementations. The generic callbacks reduce readability.

File: src/query/plan/rule_based_planner.cpp [hunk 3]
```diff
@@ -368,6 +369,12 @@ class ReturnBodyContext : public HierarchicalTreeVisitor {
 
 #undef VISIT_BINARY_OPERATOR
 
+  bool PostVisit(RangeOperator &op) override {
+    bool res = op.expr1_->Accept(*this);
+    res |= op.expr2_->Accept(*this);
```
Comment (Ignition) [inline src/query/plan/rule_based_planner.cpp:374]: Does 2nd Accept need to be visited?
Can be do a short-cutting `||`

Comment (andrejtonev) [reply src/query/plan/rule_based_planner.cpp:374]: We want to check both branches.

File: src/query/plan/preprocess.cpp (hunk 2 above)
Comment (andrejtonev) [reply src/query/plan/preprocess.cpp:190]: Sure

File: src/query/frontend/ast/pretty_print.cpp (hunk 1 above)
Comment (andrejtonev) [reply src/query/frontend/ast/pretty_print.cpp:255]: This is used when outputting the plan as a JSON.
The JSON is sent as part of the summary when running an EXPLAIN query.

Query (with index):
```
EXPLAIN MATCH (n:l) WHERE 1 < n.p < 10 RETURN n;
```
Explain query result:
```
+---------------------------------------------+
| QUERY PLAN                                  |
+---------------------------------------------+
| " * Produce {n}"                            |
| " * ScanAllByLabelPropertyRange (n :l {p})" |
| " * Once"                                   |
+---------------------------------------------+
```
JSON:
```
"explain": {
  "input":{
    "input":{
      "name":"Once"
    },
    "label":"l",
    "lower_bound":{
      "type":"exclusive",
      "value":"(ParameterLookup 7)"
    },
    "name":"ScanAllByLabelPropertyRange",
    "output_symbol":"n",
    "property":"p",
    "upper_bound":{
      "type":"exclusive",
      "value":"(ParameterLookup 13)"
    }
  }
}
```

**Comment by kgolubic** [comment]
@andrejtonev I'm missing the Release note entry.

**Comment by andrejtonev** [comment]
### Description

A query like 
`MATCH(n) WHERE n.p > 10 AND n.p < 20 RETURN n;`
will generate a tree with 2 independent filters (`n.p > 10` and `n.p < 20`).
This becomes a problem when we come to our optimizer. 
The optimizer tries to embed filters into range scans. The optimizer will embed one filter, but not the other. It limits the scans to a single filter, so only the upper or lower bound gets embedded.

This PR tries to introduce a new operator RangeOperator. 
It should fix this problem and generate a single filter with a lower and upper bound, which gets correctly inserted into a ranged scan.
The operator is not generated by the `CypherMainVisitor` but later as part of the tree rewrite. This is so other optimizations that move filters around can get picked up and potentially combine filters from different parts of the query into a single range.

[master < Task] PR
- [ ] Provide the full content or a guide for the final git message
    - **[FINAL GIT MESSAGE]**


### Documentation checklist
- [ ] Add the documentation label tag
- [ ] Add the bug / feature label tag
- [ ] Add the milestone for which this feature is intended
    - If not known, set for a later milestone
- [ ] Write a release note, including added/changed clauses
    - **Better range filter optimizations**
- [ ] Link the documentation PR here
    - **[Documentation PR link]**
- [ ] Tag someone from docs team in the comments

**Review by Ignition — APPROVED** [review]

**Comment by kgolubic** [comment]
For release note I'll use:

- Improved query optimization by unifying multiple property-based filters into a single filter that applies both upper and lower bounds. 
[#2055](https://github.com/memgraph/memgraph/pull/2055)

**Pull request merged**